Lib/test/test_sys.py

```diff
@@ -142,6 +142,123 @@ def f():
+    def test_set_exc_replaced_by_new_exception_and_restored(self):
+        exc = ValueError()
+        sys._set_exception(exc)
+        self.assertIs(sys.exception(), exc)
+        try:
+            raise TypeError()
+        except TypeError:
+            self.assertIsInstance(sys.exception(), TypeError)
+            self.assertIs(sys.exception().__context__, exc)
+
+        self.assertIs(sys.exception(), exc)
+
+    def test_set_exc_popped_on_exit_except(self):
+        exc = ValueError()
+        try:
+            raise TypeError()
+        except TypeError:
+            self.assertIsInstance(sys.exception(), TypeError)
+            sys._set_exception(exc)
+            self.assertIs(sys.exception(), exc)
+        self.assertIsNone(sys.exception())
+
+    def test_cleared_exc_overridden_and_restored(self):
+        try:
+            raise ValueError()
+        except ValueError:
+            try:
+                raise TypeError()
+            except TypeError:
+                self.assertIsInstance(sys.exception(), TypeError)
+                sys._set_exception(None)
+                self.assertIsNone(sys.exception())
+                try:
+                    raise IndexError()
+                except IndexError:
+                    self.assertIsInstance(sys.exception(), IndexError)
+                    self.assertIsNone(sys.exception().__context__)
+                self.assertIsNone(sys.exception())
+            self.assertIsInstance(sys.exception(), ValueError)
+        self.assertIsNone(sys.exception())
+
+    def test_clear_exc_in_generator(self):
+        def inner():
+            self.assertIsNone(sys.exception())
+            yield
+            self.assertIsInstance(sys.exception(), ValueError)
+            sys._set_exception(None)
+            self.assertIsNone(sys.exception())
+            yield
+            self.assertIsNone(sys.exception())
+
+        # with a single exception in exc_info stack
+        g = inner()
+        next(g)
+        try:
+            raise ValueError()
+        except:
+            self.assertIsInstance(sys.exception(), ValueError)
+            next(g)
+            self.assertIsInstance(sys.exception(), ValueError)
+        self.assertIsNone(sys.exception())
+        with self.assertRaises(StopIteration):
+            next(g)
+        self.assertIsNone(sys.exception())
+
+        # with multiple exceptions in exc_info stack by chaining generators
+        def outer():
+            g = inner()
+            self.assertIsNone(sys.exception())
+            yield next(g)
+            self.assertIsInstance(sys.exception(), TypeError)
+            try:
+                raise ValueError()
+            except:
+                self.assertIsInstance(sys.exception(), ValueError)
+                self.assertIsInstance(sys.exception().__context__, TypeError)
+                yield next(g)
+                self.assertIsInstance(sys.exception(), ValueError)
+            self.assertIsInstance(sys.exception(), TypeError)
```

Comment:
Note: This check is currently failing. The `sys.exception()` is `None` instead of the expected `TypeError`. I *think* this test is correct, so have been trying to figure out why the exception is considered cleared here but haven't been having much luck.

Reply:
Is it because there is a None on the exception stack and it's not being skipped over anymore? (Maybe we need it to be NULL rather than None?)